funding.py

```diff
@@ -184,149 +238,160 @@
-    tax_rate = agi_tax / 100
-    target_persons['new_taxes'] = target_persons.adjginc * tax_rate
+    # combine lists and initialize
+    taxes_benefits = taxes + benefits
+    spmu['new_resources'] = spmu.spmtotres
+    revenue = 0
     
-    # Calculate the total tax increase of an SPM unit
-    spmu = target_persons.groupby(['spmfamunit'])[['new_taxes']].sum()
-    spmu.columns = ['total_tax_increase']
-    target_persons = target_persons.merge(spmu,left_on=['spmfamunit'],
-                                           right_index=True)
+    # Calculate the new revenue and spmu resources from tax and benefit change
+    for tax_benefit in taxes_benefits:
+        if not ('fedtaxac' in taxes_benefits) and (tax_benefit in ['eitc', 'ctc']):
+            spmu.new_resources -= spmu[tax_benefit]
+            revenue += mdf.weighted_sum(spmu, 'tax_benefit', 'spmwt')
     
-    # Calculate funding from taxes
-    funding = (target_persons.new_taxes * target_persons.asecwt).sum()
+    # Calculate the new taxes from flat tax on AGI
+    tax_rate = agi_tax / 100
+    spmu['new_taxes'] = spmu.adjginc * tax_rate
     
-    #Calculate SPM unit new resources after taxes
-    target_persons['new_spm_resources'] = target_persons.spmtotres - target_persons.total_tax_increase
+    spmu.new_resources -= spmu.new_taxes
+    revenue += mdf.weighted_sum(spmu, 'new_taxes', 'spmwt')
     
-    if 'ssi' in benefits:
-        funding += (target_persons.incssi * target_persons.asecwt).sum()
-        target_persons.new_spm_resources -= target_persons.incssi
+    # Calculate the total UBI a spmu recieves based on exclusions
+    if ('children' in exclude) and ('non_citizens' not in exclude):
+        ubi_population = population - child_population
+        ubi = revenue / ubi_population
+        spmu['total_ubi'] = ubi * (spmu.numper - spmu.child)
     
-    if 'unemp' in benefits:
-        funding += (target_persons.incunemp * target_persons.asecwt).sum()
-        target_persons.new_spm_resources -= target_persons.incunemp
-        
-    if 'eitc' in benefits:
-        funding += (target_persons.eitcred * target_persons.asecwt).sum()
-        target_persons.new_spm_resources -= target_persons.spmeitc
-        
-    if 'ctc' in benefits:
-        funding += (target_persons.ctc * target_persons.asecwt).sum()
-        target_persons.new_spm_resources -= target_persons.spmctc
-        
-    if 'snap' in benefits:
-        funding += (target_persons.snap_pp * target_persons.asecwt).sum()
-        target_persons.new_spm_resources -= target_persons.snap_pp
-        
-    if 'energy' in benefits:
-        funding += (target_persons.energy_pp * target_persons.asecwt).sum()
-        target_persons.new_spm_resources -= target_persons.energy_pp
+    if ('children' not in exclude) and ('non_citizens' in exclude):
+        ubi_population = population - non_citizen_population
+        ubi = revenue / ubi_population
+        spmu['total_ubi'] = ubi * (spmu.numper - spmu.non_citizen)
```

Comment:
For clarity I'd suggest making a column with UBI recipients, then you can calculate the UBI (amount and column) outside the `if` statements